src/v/features/feature_state.h

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright 2022 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "cluster/types.h"
+
+namespace features {
+
+struct feature_spec;
+
+/**
+ * Feature states
+ * ==============
+ *
+ * Start as unavailable.  Become available once all nodes
+ * are recent enough.
+ *
+ * Once available, either advance straight to 'preparing'
+ * if available_policy is 'always', else wait for administrator
+ * to activate the feature.
+ *
+ * Once in preparing, either advance straight to 'active'
+ * if prepare_policy is 'always', else wait for notification
+ * that preparation is complete before proceeding.
+ *
+ * Features may be disabled at any time, but from unavailable/available
+ * states they go to disabled_clean, whereas from other states they
+ * go to disabled_dirty.  This tracks whether the feature may have
+ * written persistent structures.
+ *
+    ┌──────────────┐           ┌──────────────────┐
+    │              ├──────────►│                  │
+    │  unavailable │           │  disabled_clean  │
+    │              │◄──────────┤                  │
+    └───────┬──────┘           └───┬──────────────┘
+            │                      │            ▲
+            ▼                      │            │
+    ┌──────────────┐               │            │
+    │              │◄──────────────┘            │
+    │   available  │                            │
+    │              ├────────────────────────────┘
+    └───────┬──────┘
+            │
+            ▼
+    ┌──────────────┐            ┌─────────────────────┐
+    │              ├───────────►│                     │
+    │   preparing  │            │  disabled_preparing │
+    │              │◄───────────┤                     │
+    └───────┬──────┘            └─────────────────────┘
+            │
+            ▼
+    ┌──────────────┐            ┌─────────────────────┐
+    │              ├───────────►│                     │
+    │    active    │            │  disabled_active    │
+    │              │◄───────────┤                     │
+    └──────────────┘            └─────────────────────┘
+ *
+ */
+class feature_state {
+public:
+    enum class state : uint8_t {
```

Comment:
unless the explicitly underlying type here is used for compatibility reasons (e.g. adl stuff) then it's unnecessary in serde where every enum is int32_t

Reply:
I was just going for efficiency (perhaps prematurely), using one byte when an enum looks like it's going to never have more than a handful of entries.

Are you saying that serde will encode this as a uint32_t even if the underlying type is uint8_t though?

Reply:
> Are you saying that serde will encode this as a uint32_t even if the underlying type is uint8_t though?

yes (well, an int32_t). when efficiency of size is important (e.g. packing in many such instances) then one can opt-out and customize on a per enum-class type.